packages/workspace/src/browser/workspace-service.ts

```diff
@@ -88,71 +89,71 @@ export class WorkspaceService implements FrontendApplicationContribution {
+        if (configUri) {
+            this.watcher.watchFileChanges(configUri);
+        }
+    }
 
-            for (const rootBeingWatched of Object.keys(this.rootWatchers)) {
-                if (roots.indexOf(rootBeingWatched) < 0) {
-                    this.stopWatch(rootBeingWatched);
-                }
-            }
-            this._roots.length = 0;
-            for (const rootToWatch of roots) {
-                const valid = await this.toValidRoot(rootToWatch);
-                if (!this.rootWatchers[rootToWatch]) {
-                    await this.startWatch(valid);
-                }
-                if (valid) {
-                    this._roots.push(valid);
-                }
-            }
-            if (!this.rootWatchers[this._workspaceFolder.uri]) {
-                // must watch the workspace folder for meta data changes, even if it is not in the workspace
-                await this.startWatch(this._workspaceFolder);
+    protected async updateRoots(): Promise<void> {
+        this._roots = await this.computeRoots();
+        this.deferredRoots.resolve(this._roots); // in order to resolve first
+        this.deferredRoots = new Deferred<FileStat[]>();
+        this.deferredRoots.resolve(this._roots);
+        this.onWorkspaceChangeEmitter.fire(this._roots);
+    }
+
+    protected async computeRoots(): Promise<FileStat[]> {
+        const config = await this.getWorkspaceConfig();
+        if (!config || !config.roots.length) {
+            return this._workspaceFolder ? [this._workspaceFolder] : [];
+        }
+        const roots: FileStat[] = [];
+        for (const rootUri of config.roots) {
+            const root = await this.toValidRoot(rootUri);
+            if (root) {
+                roots.push(root);
             }
         }
-        this.onWorkspaceChangeEmitter.fire(this._roots);
+        return roots;
     }
 
-    protected async getRootConfig(): Promise<{ stat: FileStat | undefined, roots: string[] }> {
+    protected async getWorkspaceConfig(): Promise<{ stat: FileStat, roots: string[] } | undefined> {
+        if (!this.preferences['workspace.supportMultiRootWorkspace']) {
+            return undefined;
+        }
         const configUri = this.getWorkspaceConfigFileUri();
         if (configUri) {
-            let fileStat = undefined;
             const uriStr = configUri.path.toString();
             if (await this.fileSystem.exists(uriStr)) {
                 const { stat, content } = await this.fileSystem.resolveContent(uriStr);
-                fileStat = stat;
                 if (content) {
-                    const roots = JSON.parse(content).roots || [];
-                    return { stat, roots: this._workspaceFolder && roots.length === 0 ? [this._workspaceFolder.uri] : roots };
+                    // FIXME use jsonc + ajx to parse & validate content
+                    const roots: string[] = JSON.parse(content).roots || [];
+                    return { stat, roots };
```

Comment:
in the case that `roots == [ ]`, all watchers are disposed - at this point if the user adds a new root to the workspace, s/he wouldn't see that new root, because the workspace config file is not being watched anymore. 

but it is ok - my PR in review makes the config file independent from the root folders, so it is not going to be an issue.

Reply:
ahh, you are right !